plutus-metatheory/src/VerifiedCompilation/UCaseOfCase.lagda.md

```diff
@@ -0,0 +1,100 @@
+```
+module VerifiedCompilation.UCaseOfCase where
+
+```
+
+## Imports
+
+```
+open import Data.Nat using (ℕ;zero;suc)
+open import Data.Empty using (⊥)
+open import Scoped using (ScopedTm; Weirdℕ)
+open import Untyped using (_⊢; case; builtin; _·_; force; `; ƛ; delay; con; constr; error)
+open import Untyped.CEK using (BApp; fullyAppliedBuiltin; BUILTIN; stepper; State; Stack)
+open import Evaluator.Base using (maxsteps)
+open import Builtin using (Builtin; ifThenElse)
+open import Data.List using (List; zip; [_])
+open import Utils as U using (Maybe;nothing;just;Either)
+open import RawU using (TmCon)
+open import Data.List.Relation.Unary.All using (All; []; _∷_)
+open import Data.Product using (_×_; proj₁; proj₂)
+open import Relation.Nullary using (Dec; yes; no)
+```
+
+## Translation Relation
+
+This compiler stage only applies to the very specific case where an `IfThenElse` builtin exists in a `case` expression.
+It moves the `IfThenElse` outside and creates two `case` expressions with each of the possible lists of cases. 
+
+This translation relation has a constructor for the specific case, and then inductive constructors for everything else
+to traverse the ASTs.
+
+```
+data _⊢̂_⊳̂_ (X : Set) : (X ⊢) → (X ⊢) → Set where
+--   refl : ∀ {X : Set} {x : X ⊢} → X ⊢̂ x ⊳̂ x
+   caseofcase : ∀ {b : X ⊢} {tn tn' fn fn' : ℕ} {alts alts' tt ft tt' ft' : List (X ⊢)} 
+                → All (λ altpair → X ⊢̂ (proj₁ altpair) ⊳̂ (proj₂ altpair)) (zip alts alts') -- recursive translation for the other case patterns 
```

Comment:
What if `alts` and `alt'` have different lengths? Same question about `tt`/`tt'` and `ft`/`ft'`.

Also I'm very much not a fan of the cryptic names.

Reply:
If the lengths differ then the two things shouldn't match, but the Pointwise approach probably handles this more gracefully.

Reply:
They shouldn't match, but they will, will they not? This is how `zip` is defined:

```
zipWith : (A → B → C) → List A → List B → List C
zipWith f (x ∷ xs) (y ∷ ys) = f x y ∷ zipWith f xs ys
zipWith f _        _        = []
```

it drops all the extra elements of the longer list. And then `All` will not apply to them. And so you can introduce arbitrary garbage at the end of one of the lists and it will be ignored by the translation.

Or am I missing something?

Reply:
This is a good point - to be honest, I put the zip in as a placeholder because I didn't like it but hadn't thought of a better solution. I'll look further at the Pointwise stuff.